src/raw_rtde/connection.rs

```diff
@@ -0,0 +1,891 @@
+        Self::decode_controller_version_reply(reply)
+    }
+
+    /// Negotiates the protocol version.
+    ///
+    /// Needs to be done before sending other packets. `RawRTDE::connect` does it automatically.
+    pub fn negotiate_protocol_version(&mut self) -> Result<(), String> {
+        let request = Self::encode_protocol_version_request(RTDE_VERSION)?;
+        self.send_packet(&request)?;
+        let reply = self.wait_for_reply(RTDECommand::RequestProtocolVersion as u8)?;
+        let protocol_ok = Self::decode_protocol_version_reply(reply)?;
+
+        match protocol_ok {
+            true => Ok(()),
+            false => Err(format!(
+                "Robot did not accept RTDE protocol version {}",
+                RTDE_VERSION
+            )),
+        }
+    }
+
+    fn encode_protocol_version_request(version: u16) -> Result<Vec<u8>, String> {
+        let mut payload: Vec<u8> = vec![];
+        payload.write_u16::<BigEndian>(version).unwrap();
+        Self::encode_packet(RTDECommand::RequestProtocolVersion as u8, payload)
+    }
+
+    fn decode_protocol_version_reply(payload: Vec<u8>) -> Result<bool, String> {
+        let mut rdr = Cursor::new(payload);
+        let protocol_ok = rdr
+            .read_u8()
+            .map_err(|_e| "Protocol version reply too short")?
+            != 0;
+        Ok(protocol_ok)
+    }
+
+    fn encode_controller_version_request() -> Result<Vec<u8>, String> {
+        Self::encode_packet(RTDECommand::GetControllerVersion as u8, vec![])
+    }
+
+    fn decode_controller_version_reply(payload: Vec<u8>) -> Result<ControllerVersion, String> {
+        let mut rdr = Cursor::new(payload);
+        let major = rdr
+            .read_i32::<BigEndian>()
+            .map_err(|_e| "Controller version reply too short")?;
+        let minor = rdr
+            .read_i32::<BigEndian>()
+            .map_err(|_e| "Controller version reply too short")?;
+        let bugfix = rdr
+            .read_i32::<BigEndian>()
+            .map_err(|_e| "Controller version reply too short")?;
+        let build = rdr
+            .read_i32::<BigEndian>()
+            .map_err(|_e| "Controller version reply too short")?;
+        Ok(ControllerVersion {
+            major,
+            minor,
+            bugfix,
+            build,
+        })
+    }
+
+    fn encode_payloadless_request(command: RTDECommand) -> Result<Vec<u8>, String> {
+        Self::encode_packet(command as u8, vec![])
+    }
+
+    fn decode_status_only_reply(payload: Vec<u8>) -> Result<bool, String> {
+        let mut rdr = Cursor::new(payload);
+        let accepted = rdr
+            .read_u8()
+            .map_err(|_e| "Controller version reply too short")?;
+        Ok(accepted != 0)
+    }
+
+    fn encode_configure_inputs_request<U: AsRef<str>>(variables: &[U]) -> Result<Vec<u8>, String> {
+        let variables = variables.iter().map(|v| v.as_ref()).collect::<Vec<&str>>();
+        if !variables.iter().all(|v| !v.contains(',')) {
```

Comment:
That's a lot of negative logic. This would be clearer as `variables.iter().any(|v| v.contains(','))`.